kong/core/handler.lua

```diff
@@ -118,16 +125,10 @@ return {
           "' with: "..tostring(err))
       end
 
-      -- if set `host_header` is the original header to be preserved
-      var.upstream_host = host_header or 
+      -- if set, `upstream_host` is the original header to be preserved
+      var.upstream_host = ctx.upstream_host or
           balancer_address.hostname..":"..balancer_address.port
 
```

Comment:
I think most of this block, where we assign the results back to the nginx core, should move to the balancer context. Because in the balancer context is where we do the retries if something fails, and it might alter `balancer_address.hostname..":"..balancer_address.port` results